src/objectid.ts

```diff
@@ -103,17 +133,38 @@ export class ObjectId extends BSONValue {
       workingId = inputId;
     }
 
+    // If we have reached the end of the pool then create a new pool
+    if (!pool || poolOffset + 12 > pool.byteLength) {
+      createPool();
+    }
+    this[_pool] = pool;
+    this[_offset] = poolOffset;
+    poolOffset += 12;
+
     // The following cases use workingId to construct an ObjectId
     if (workingId == null || typeof workingId === 'number') {
       // The most common use case (blank id, new objectId instance)
       // Generate a new id
-      this.buffer = ObjectId.generate(typeof workingId === 'number' ? workingId : undefined);
-    } else if (ArrayBuffer.isView(workingId) && workingId.byteLength === 12) {
-      // If intstanceof matches we can escape calling ensure buffer in Node.js environments
-      this.buffer = ByteUtils.toLocalBufferType(workingId);
+      ObjectId.generate(
+        typeof workingId === 'number' ? workingId : undefined,
+        this[_pool],
+        this[_offset]
+      );
+    } else if (ArrayBuffer.isView(workingId)) {
+      if (workingId.byteLength !== 12 && typeof inputIndex !== 'number') {
+        throw new BSONError('Buffer length must be 12 or offset must be specified');
+      }
+      if (
+        inputIndex &&
+        (typeof inputIndex !== 'number' || inputIndex < 0 || workingId.byteLength < inputIndex + 12)
```

Comment:
Went down a bit of a rabbit hole analyzing this condition. 😅 It stood out to me that anything falsy won't make it to the second part that does a type and bounds check. Turns out if an empty string, or false are passed in here then the next part still works `'' + 1` will still get index `1` as will `false + 1`. There are two more falsy things that will cause an issue, the easy one is a bigint: `0n` typescript would prevent that but we can also elminate that with the type check moved before this. The other is `NaN`, which _should_ never happen, but the bounds check would elimnate it if was able to reach that check.

Reply:
I really appreciate the thorough review 😄 Took me a little while to wrap my head around this but I think I've improved the check here and I've added more tests to cover these possibilities. 

- Be more explicit when byteLength === 12 and ignore inputIndex entirely
- No falsy check
- Add isNaN

Let me know if I've missed anything here.

Reply:
Looks great! This bit is internal so its use is "conrolled" but I think it's good we made it obvious to the next maintainer if a slight math operation goes wrong and they get NaN it will crash loudly.